### ISSUE TITLE
RFE: output non-latin text as text instead of glyphs when printing

### ISSUE THREAD

**Issue by Vasja1986** [taviso/wpunix#7]
WordPerfect currently outputs postscript drawing instructions for each and every single (non-latin) glyph when printing instead of the actual text. Thus when printing to PDF the text in the resulting file is not selectable, because it is just a bunch of images. Even worse the glyphs that WordPerfect draws are very poor in terms of quality. Some of the Cyrillic glyphs are barely legible.

It would be great if:
- a postscript encoding table for each and every WP charset gets included in the postscript output;
- the corresponding encoding table gets selected when changing WP charset;
- the text would be outputted in strings of text instead of drawing instructions for individual glyphs.

**Comment by emendelson** [comment]
@Vasja1986 - This is the effect of the printer driver, not of WordPerfect itself, and the only way to change it is to use a PostScript driver that supports non-latin glyphs. You might look at the WPDOS 5.1 drivers on my WPDOS site, which are fully compatible with WP Unix:

http://wpdos.org/51printers.html

Also, you may need to make use of the Greek and Cyrillic PostScript driver that Corel and WP distributed on their language-pack CD and as separate packages. I have copies of those if you want to work with them.

**Comment by taviso** [comment]
I wonder if I should include your driver in the package, is there any reason someone might not want it? If there are no downsides, it seems like we should just include it!

**Comment by emendelson** [comment]
@taviso - WP printer drivers are stored in .ALL files that contain definitions for one or more printer types. When you select a printer in WP, WP creates a .PRS file by extracting data from the .ALL file. 

In the WP 5 era, an ALL file could contain a set of PostScript definitions, or a set of other types of definitions (HP PCL, Epson Esc-E, etc.). I'll put together two ALL files: one for PostScript, with the Latin, Greek, and Cyrllic definitions, and one for HP PCL (with a similar set of definitions). You could certainly add those - as there are no downsides.

I'll look at the Greek/Cyrillic drivers later on to see whether it's possible to combine them. But it may take your PostScript skills to make a single driver. I know how to manipulate PostScript fonts but not ordinary PostScript code beyond a few things I fixed in the old drivers. And I don't know how modern PostScript printers support multiple alphabets. If the glyph is present in the hardware, then a WP driver can use it - in the way that my page explains how to add Euro support to an existing hardware driver:

http://www.columbia.edu/~em36/wpdos/typeinstall.html

As for the rest of the utf8 set, that may be a problem for the future.

**Comment by emendelson** [comment]
To continue this: I've made a start on this by creating a printer definition that uses a PostScript Type 1 soft font that contains a large number of glyphs. I've added three or four Greek characters to the WP driver so that those characters print as text. It would take an enormous amount of work to complete this, but it's possible to do. More details later.

**Comment by emendelson** [comment]
@Vasja1986 - Here is a demonstration that what you're asking for can be done, though it will take a lot of effort to finish the job.

Download this: 

https://www.dropbox.com/s/b3kmiup90sj0vnq/WP-UTF8.zip?dl=1

It contains an .all file and a .pfb file. Make sure the names are all lower-case. Sudo-move the .all file into /opt/wp70/shlib10; move the .pfb file into ~/.wprc

In WP, select an additional printer, choose `PostScript utf8 SoftFont (woods.org)`. Make the destination DISK, but rename the destination file from <doc>.dsk to <doc>.ps

In the Edit printer screen, set the default font to Monospace (any size is OK). Specify the directory for soft fonts, etc. as `/home/<yourusername>/.wprc` (though `~/.wprc` should work also, I think).

Print some text to make sure it's working, and convert the ps output to PDF (if GhostScript is installed, use `ps2pdf`.)

The printer driver contains the codes for the first seven characters in WP character set 8 (Greek) and the first seven characters in WP set 10 (Cyrillic). You can use the printer definition program (PTR.EXE) under DOS to explore how this is done. Go to Fonts, Monospace, Character set, and look at sets 8 and 10 for details.

The font is a PostScript Type 1 font exported from FontLab 7 from Microsoft's freely-available TrueType version of Courier. It contains everything in UTF8, I believe, but it would take hours of work to create the printer driver Some of it could be automated with a WP Shell macro, but that require the WP Shell running under DOS - unless we can find WordPerfect Office for Unix with the Shell and the Unix printer definition program.

@Vasja1986 - If this is important to you, I'll be glad to give any advice that will help you get the job done. It will be infinitely easier to do this with monospace fonts than proportional ones, but that may be all that anyone needs. And it doesn't have to be done separately for bold, italic, and bold italic. When you create a character map for one monospace font, it will work perfectly with any other monospace font that uses the same character names.

**Comment by Vasja1986** [comment]
@emendelson I have downloaded the files from your dropbox link and can confirm that this proof of concept works. It looks very promising. Your posts are quite extensive and to my regret I have to admit that due to a short supply of time I am unable at the moment to go into much detail in my response, although you raise a number of valid points worth addressing. So this is basically a short note to show you appreciation for your hard work. I hope we can find a way to automatically provide input to the PTR program in order to avoid a tedious labour intensive process. Maybe I can explore some possibilities at a later time. Meanwhile I keep monitoring this ticket for updates from you and @taviso.

**Comment by Vasja1986** [comment]
By the way I do think that we need a single unified printer driver. It is then up to CUPS to translate the output in appropriate commands for each printer. On modern linux there is no need to bypass CUPS and provide direct output to a printer.

**Comment by emendelson** [comment]
@Vasja1986 - If you are willing to do the work of writing a unified printer driver that would talk to CUPS, that would be wonderful. No one else is likely to do it, but it may be possible. As you know WP expects to send raw data to the printer, but perhaps it would be possible to overcome it.

Again, if this is something that you want, I hope you will get to work on it as soon as possible. it would be very good to have.

As for automating the creation of a printer driver: Many years ago, I used Shell macros to copy character widths from WPDOS 6.x drivers to WPDOS 5.x drivers (calculating the different widths required). That worked well when using the Shell to transfer data from two different WP programs. It will be more difficult to do this with the contents of a soft font, as one would need some way to get the glyphname from the font and match it to the data in the Printer Definition Program. Not impossible, but a lot of work. 

Again, if you're willing to do the work - and you're not merely asking someone else to do the hard work for you - I'll be glad to provide pointers. If you have no intention of doing any of the work on your own, then I won't take up more of your time with this discussion.

**Comment by taviso** [comment]
I see, so I think we need to:

1. Convert all the wordperfect characters to postscript glyphs.
2. Find the widths for every character in the proportional fonts.
3. Either write a tool to generate the all files, or script this ptr tool.

Well, 1 seems pretty easy, we can just reuse the `libwpd` tables. I know PostScript, but I don't know much about typography. Looking at the tables it already has, it thinks the letter `K` has width 722 in Times New Roman at point size 12.

I dunno what unit "722" is, 1/1000ths of an inch? seems such a weird unit.

Well, whatever, I can ask ghostscript how big it thinks that character is and just find the correct divisor to translate them?

```
GS>/Times-Roman findfont 12 scalefont setfont
Loading NimbusRoman-Regular font from /usr/share/ghostscript/9.50/Resource/Font/NimbusRoman-Regular... 4377864 2873199 2752200 1455440 1 done.
GS>(K) stringwidth pop 722 exch div ==
83.3335571
```

Okay, so let's try `83.3335571`

```
GS>(H) stringwidth pop 83.3335571 mul cvi =
722
GS>(I) stringwidth pop 83.3335571 mul cvi =
333
GS>(J) stringwidth pop 83.3335571 mul cvi =
388
```

![image](https://user-images.githubusercontent.com/123814/182755667-4b591293-346c-4c22-96dc-a13edc731123.png)

Hey, that looks pretty good!

Okay, so we can autogenerate the widths and glyphs, we just need to get them into the .all file.

Let me read the developer SDK and see if I should just writes a tool to generate these files, or if it's easier to script this PTR utility. As this only has to be done once, I'm leaning towards just some one-off ugly scripting rather than investing any time in writing a new tool.

**Comment by emendelson** [comment]
This is amazing. I think there are ways to generate those files, but haven't looked at the SDK for a while. There's a third-party utility that takes the data from an .AFM file (used for installing PostScript Type 1 fonts) and feeds it into a printer driver:

http://www.columbia.edu/~em36/wpdos/typeinstall.html#installer

It's easy to create AFM files from Type 1 fonts (FontForge should do it) - but I'm NOT sure that AFM files include the metrics for non-encoded characters, and all the characters that we're adding won't be encoded in the font. (If any of this isn't clear, just shout.)

Also, yes 1000/inch is the PostScript Type 1 font unit of measurement. Those were early days. The first authors, Warnock and Geschke, were basically inventing everything as they went along. The PostScript (and I think TrueType) name for the French quotation marks (guillemot) is misspelled guillemet because they mistyped the correct name.

**Comment by taviso** [comment]
Thanks, the addall utility seems promising. Absolutely baffling decision to make it use a binary command format and not plain text, but at least it's a relatively simple structure.

Okay, I'll have a go at generating these operation files.

**Comment by emendelson** [comment]
When mapping symbols in a font to the WP character set, would it be possible to do something like this:

If the glyph name begins `uni` followed by four text digits, then look up the hex digits in that table of unicode equivalents (from the libwpd project) and set the WP character number according to the unicode number's place in the table - e.g. if the glyph name is uni0666 (I'm inventing this example), then look for 0x0666 in the libwpd table, make sure you're using the WP5 version, and assign the glyph to whatever location is in the table, e.g., 10,107 (again, completely invented).

What I don't think exists is a table that matches Type 1 glyph names like Beta and beta to WP character codes like 8,3 and 8,4. You've probably already thought of this method of making such a list. Use an Adobe glyph table like this:

https://en.wikipedia.org/wiki/Adobe_Glyph_List

For each named character in the glyph list get the unicode number from the first column, find that number in the libwpd table, and add the glyph name and the WP code to the list.

Incidentally, any work that gets done on a WP 5-style printer driver like this can be used by WP 6.x or later, because the WP 6 version of PTR.EXE can convert WP 5 drivers into WP 6+ format (but can't go in the reverse direction: you can't create a WP 5 driver from a WP 6 version).

Apologies if you've already figured out all of this, as you probably have.

**Comment by taviso** [comment]
Mapping the glyphnames is no problem, it's getting that data into the structure that addall wants that's the roadblock. Well, it's not really a roadblock, just some tedious work translating their documentation into the code to generate it!

**Comment by emendelson** [comment]
... and I finally saw that PTINSTAL.INI contains mapping for all the glyph names that were current at the time... So much for my elaborate schemes for doing exactly that...

**Comment by taviso** [comment]
I experimented with it last night, I've got some prototype code working. I think after all this work to generate the tables that `addall` wants, I might as well just generate the `.all` file myself -- it's just a marginal increase in work at this point, and then there is no dependency on DOS.

This has other advantages, I can just query ghostscript about all the fonts it knows about, and generate them dynamically, so any system font you have installed will work in WordPerfect.

I assume this will work in Windows too, you can just `ln -s /mnt/c/Windows/Fonts/ ~/.fonts` in WSL, then if the command `fc-list` can see them then I think ghostscript will see it.

The end result should be that you can just run one command and any system fonts will be visible to WordPerfect. At least, that's the theory...

Anyway, this isn't a small job, but I'll hack on it this weekend.

**Comment by emendelson** [comment]
> I experimented with it last night, I've got some prototype code working. I think after all this work to generate the tables that `addall` wants, I might as well just generate the `.all` file myself -- it's just a marginal increase in work at this point, and then there is no dependency on DOS.
> 
> This has other advantages, I can just query ghostscript about all the fonts it knows about, and generate them dynamically, so any system font you have installed will work in WordPerfect.
> 
> I assume this will work in Windows too, you can just `ln -s /mnt/c/Windows/Fonts/ ~/.fonts` in WSL, then if the command `fc-list` can see them then I think ghostscript will see it.
> 
This will be amazing. character-mode WP Unix had no support for TTF or OTF fonts - will this make it possible to use those fonts with WP??

**Comment by taviso** [comment]
> This will be amazing. character-mode WP Unix had no support for TTF or OTF fonts - will this make it possible to use those fonts with WP??

Yep, should be no problem. The one caveat will be that if you add a new system font (on Linux you typically just drop the .ttf file into ~/.fonts) you will have to regenerate the .all file before WP knows about it.

I suppose we could patch wp to do this on startup, but I'll cross that bridge when I get to it!

**Comment by emendelson** [comment]
It would be good to be able to limit the font list in WP to something that won't require endless scrolling That might be doable by the individual user from the Select Printer, Edit Printer Setup, Cartridges/Fonts etc menu, but it would be good to have ptr in order to edit the PRS file. I know I'm asking for the moon...!

**Comment by Vasja1986** [comment]
Wow! This is amazing. I definitely did not expect such a comprehensive solution in such a sort time-frame, when I opened this ticket. It almost feels like getting a Christmas present in the middle of summer. Many thanks.

I do have a question though: when adding all the system fonts to the list, you are obviously going to run into some fonts that do not contain all national language characters. How are you going to handle such cases? Programmatically, it seems to me, that the easiest thing one can do is setting the width of missing characters to zero. However this is not very desirable. It would be better if some placeholder character (question mark or box drawing character) substitutes the missing characters, so the user would get a clue about the problem. Even better would be to use a fallback font in such cases. But maybe that will be very complex to implement.

**Comment by emendelson** [comment]
@Vasja1986 - WordPerfect is very intelligent about characters that aren't in a  font. If the character width is not specified in the driver - and if the character doesn't exist in the font, then there will be no character width in the driver - then WP will either use a substitute font that contains the character, or if no suitable font is available, it will print a graphic glyph. This is all explained in the manual in the sections about Font Substation. There is nothing special that needs to be done to make this work correctly.

**Comment by taviso** [comment]
I have code that can generate input for addall from a font (or .ttf), I haven't written the code to make the all file myself yet.

This works okay, but I didn't realize that postscript doesn't natively handle kerning, it just gives you a callback and you're supposed to do the intra-character spacing yourself. That sucks, so I have no way of generating kerning tables.

I'm aware ghostscript can generate .afm files, and .afm files *can* contain kerning pairs, but the ones that ghostscript generates doesn't. Presumably this is because PostScript is oblivious to kerning so why would it?

I guess I don't know what the answer is, require that you generate an afm file to use a font? You don't **have** to, but then you will have no kerning. The lack of kerning doesn't bother me, but I know lots of people are fastidious about it (is that the right word?). I think it will likely be a big issue for some people.

There is a ttf2afm utility that comes with pdftex that works great, that would be good enough. I don't know how to do it for type1 fonts, probably fontforge?

**Comment by emendelson** [comment]
Hmmm... an interesting problem. If I remember correctly, commercial Type 1 fonts had kerning built in to the font data by the original designer, and that information went into the AFM file, and WP used it when indicating the horizontal position of each character in PostScript output. It's odd that Ghostscript doesn't output kerning pairs, but, as you say, PostScript itself doesn't know about it - but an application like WP can use kerning data. 

If the PFB file contains kerning data, then FontForge will generate an AFM with the kerning data - at least it did that when I tried it just now. But this seems to be getting very complicated...

**Comment by taviso** [comment]
I got kerning working if you supply the afm file. I just need to add monospace support, and add more error checking, which should be easy.

```
$ ./afm2add TimesNewRomanPSMT.afm 
afm2add: The afm file was not *fully* parsed (probably still okay).
kerning: begin, size@12pt
kerning: 3465 kerning pairs in TimesNewRomanPSMT.
kerning: attempting to sort pairs (slow)
kerning: sorting complete, now counting pairs
kerning: Gamma+Alpha adjusted -154? That seems a lot.
kerning: Upsilon+Alpha adjusted -164? That seems a lot.
kerning: Upsilontonos+Alpha adjusted -164? That seems a lot.
kerning: Upsilondieresis+Alpha adjusted -164? That seems a lot.
kerning: Gamma+Lambda adjusted -154? That seems a lot.
kerning: Upsilon+Lambda adjusted -164? That seems a lot.
kerning: Upsilontonos+Lambda adjusted -164? That seems a lot.
kerning: Upsilondieresis+Lambda adjusted -164? That seems a lot.
kerning: Alpha+Upsilon adjusted -151? That seems a lot.
kerning: Lambda+Upsilon adjusted -151? That seems a lot.
kerning: Alphatonos+Upsilon adjusted -151? That seems a lot.
kerning: Alpha+Upsilondieresis adjusted -151? That seems a lot.
kerning: Lambda+Upsilondieresis adjusted -151? That seems a lot.
kerning: Alphatonos+Upsilondieresis adjusted -151? That seems a lot.
kerning: 345/3465 pairs recognized
kerning: 3047 duplicate pairs eliminated
kerning: 104 ascii pairs, 241 glyph/mixed pairs
kerning: table TimesNewRomanPSMT generated
spacing: begin, size@12pt
spacing: 4786 characters in TimesNewRomanPSMT.
spacing: set 0, size: 127
spacing: set 1, size: 242
spacing: set 2, size: 28
spacing: set 3, size: 88
spacing: set 4, size: 102
spacing: set 5, size: 35
spacing: set 6, size: 238
spacing: set 7, size: 229
spacing: set 8, size: 210
spacing: set 9, size: 119
spacing: set 10, size: 150
spacing: set 11, size: 185
spacing: set 12, size: 0
spacing: recognized 523 chars
typeface: begin TimesNewRomanPSMT, capheight 662
typeface: complete
charmap: begin TimesNewRomanPSMT.
charmap: glyph count 1031
operations: begin, size@12pt
operations: Printer GhostScript
```
![image](https://user-images.githubusercontent.com/123814/183327350-2fe876b9-b336-49bc-be48-7ba9610ecf38.png)


I'll make sure it works with all the fonts that come with GhostScript, and then push it to the repo. I'm sure it will crash and/or fail on some fonts, but I will consider support for all the GhostScript standard fonts a good enough prototype!

**Comment by emendelson** [comment]
This is amazing. What is the command that you use for printing?

**Comment by taviso** [comment]
I don't really know the best way to do it yet! What would you suggest?

I'm still ironing out a few issues, for example, I was using the libwpd tables, but there are a few issues with them.

libwpd thinks a lot of the characters in Set 1 are combining characters, for example It thinks 1,2 (tilde) is U+0303. That's a combining tilde that has no width, but the the PSR files that come with WordPerfect think 1,2 has a width of around 330 postscript units, the same for grave and so on.

There is a small non-combining tilde at U+02DC (˜, small tilde), which has a width of 333, exactly the right width.. is it a mistake, and it should be that one?

The issue is, maybe they were *supposed* to be combining characters, but printers in the 90s didn't support that or only did with overstrike hacks? I don't know the answer.

It also thinks the the euro currency symbol (4,60) should be U+20A0 (₠), the old ECU symbol. I'm pretty sure it should be U+20AC (€).

**Comment by taviso** [comment]
I think I'll just get it working for now and I'll iron out the character issues later.

I'm already sick of dealing with fonts 😂

**Comment by emendelson** [comment]
Quick reply: I think libwpd was designed mostly for use in LibreOffice, and I'm not surprised it gets some things wrong. I'm looking at the Apple LaserWriter IINTX driver in WPPS1.ALL, at the Times Roman (TIR) Proportional Spacing Table, and yes, 1,2 (tilde) is 333 units wide, and is not the combining character. WP *did* use overstrike hacks when necessary, so I think you should trust the original PRS files, NOT libwpd.

If (as I think we should) we're conforming to WP for Windows, then we should be using 4,72 as the euro symbol. The original character map had circle-U there. The old ECU symbol should stay where it was, where nobody ever uses it, but it has some function in WP tables that I forget. To be clear: 4,72 should be mapped to the unicode euro, but no other currency symbol needs to be changed.

More later if needed. I saw earlier that the Ghostscript fonts have 845 characters, which should be more than enough for most purposes...

**Comment by taviso** [comment]
Okay, from within postscript you can just `/pdfwrite finddevice setdevice`, so I guess I will just do that.  A pdf file should just appear (Just like postscript passthru, but a pdf will appear).

I think I agree it makes sense to remove the combining characters, I suspect whoever made these tables just misinterpreted the intent.

It's pretty close to being usable. I think it should generate nice clean modern PDFs, and support any font you can make an AFM file for.

**Comment by emendelson** [comment]
Here's a script I used for printing to the default printer in macOS. I do something more complicated now, but I think this got the job done:

```
/usr/bin/pstopdf "$1" -o temppdf.pdf 
/usr/bin/lpr -r temppdf.pdf
rm "$1"
```
Of course the paths are macOS paths...

**Comment by taviso** [comment]
oh I see, you want wp to do the conversion automatically rather than you run `ps2pdf` or `gs` manually? Hmm, I don't know.

We can easily patch that functionality in if you feel strongly about it, I didn't really consider it, I'm pretty happy just running `ps2pdf` 😄 

You can take a look at the current progress if you like, here is a sample `.ALL` file. I'm aware there are lots of bugs and glyphs not quite right, but I think uppercase cyrillic alphabet works.

[gscript.zip](https://github.com/taviso/wpunix/files/9285757/gscript.zip)

It was all automatically generated from the AFM files in `/usr/share/fonts/type1/gsfonts/`

**Comment by emendelson** [comment]
Further to the above: The driver (which is basically amazing) has a few problems that I hope won't be hard to fix.

Many of the glyphs that are included in the character maps don't have corresponding widths, so they don't print. For example, many of the characters in sets 6 and 8, and all the Cyrllic lower-case letters in set 10.

Also, at some point, it may help to make the typefaces of the PostScript 13 be the typefaces of the URW fonts. WP uses the "typeface" setting to decide which font to use (not the name of the font), so it may choose one of the original PS13 fonts instead of the URW ones. But that's an issue for later, and may not be an issue at all.

I hope I haven't taken many more hours of your life with this report!

**Comment by taviso** [comment]
No problem, the hard part was figuring out the file format, and that's done now :)

I think I don't have much faith in the libwpd tables, and they're also not a very convenient format to edit or browse for mistakes. I'll fix that, and then the changes will automatically apply to any other fonts.

I'll try to iron out as many of these issues as I can, and then see how it looks. I suspect even if it isn't perfect it will be an improvement over using the low-quality vector characters, so I should make a new release with this driver included.

People can also add their own .ttf fonts. It's not as easy as I hoped (because you need the afm file), but at least it's possible.

**Comment by emendelson** [comment]
For what it's worth (and don't feel *any* obligation to include this in your deb file), here's a wpdest.def file that can be copied into /opt/wp70/shlib10. It provides three destinations for printing one that sends the raw printer data to the default printer, one that uses ps2pdf to send Ghostscript output to a PDF and then to the default printer, and one that writes to CUPS-PDF. All require Ghostscript to be installed; the third requires CUPS-PDF. 

The third of these is only needed on WSL, because CUPS-PDF is available as a destination in standard (non-WSL) installations.

Again, feel free to ignore this. It might be best if I simply post it on my page.

[wpdest.zip](https://github.com/taviso/wpunix/files/9294090/wpdest.zip)

**Comment by emendelson** [comment]
A few more notes on the .ALL file. First, here's a map of the WP Character Set 10 (Cyrillic) matched to the glyph names in the Ghostscript fonts. I don't know if it's useful or not. 

It occurs to me that it's possible to export an existing character map from PTR.exe, edit the map, and then import it again. The PTR menu doesn't show this, but you can export/import character maps by using Ctrl-F5, the standard export/import keystroke in WP. Anyway, here's a table for set 10:

[WPMap10.txt](https://github.com/taviso/wpunix/files/9303940/WPMap10.txt)

Next, the existing GSCRIPT.ALL is amazing, but, as you said, needs some cleanup. 4,72 should be mapped to `"/Euro glpyhshow "` (not /euro) to work correctly. The misspelling may be why there's no entry in the Proportional Spacing Tables for 4,72; I've added this entry for the Century Schoolbook L Roman font to show how it can work correctly, in this slightly revised version of the ALL file:

[gscript-revised.zip](https://github.com/taviso/wpunix/files/9303985/gscript-revised.zip)

Also, for the PostScript 13 to work correctly, the Typeface, Character Map, and Proportional Spacing Tables should be set to the Nimbus fonts, not the original fonts, so that they'll use the full character set and spacing table.

A tip and trick for the ALL files: All the proportional Ghostscript fonts (and I think the monospaced fonts but I'm not certain) use the same character set, so it's possible to rename one of the existing character sets to something like Ghostscript Proportional, and then assign that character set to all the fonts. This lets you reduce the size of the ALL file and makes corrections and fixes easier - you only have to edit one character set.

The WP shell makes it possible to automate a lot of things by running PTR under the shell and recording a macro that (for example) assigns a character set to multiple fonts. I'm happy to do any of that work if needed. Meanwhile, this is already an amazing success.

P.S. I created that map of WP Set 10 to add to the PSINSTAL.INI file for the PTINSTAL program that I posted earlier, but the program ignored the Cyrillic glyphs when importing fonts, so it's possible that I need to study that program further. I looked at the file format for ADDALL, and it only made my eyes water...

**Comment by taviso** [comment]
You can try this updated version, I think some issues are fixed.

I don't think I understand how to calculate the slant degree correctly, but maybe it's right.

I'm pretty sure the spacing is correct, and most of the glyphs work! I'll work on it some more this weekend.

[gscript.zip](https://github.com/taviso/wpunix/files/9305382/gscript.zip)

There's no issue with additional work per table, it's all automated - when one is fixed, they all get fixed!

**Comment by taviso** [comment]
Oh - I've noticed the kerning works, but is not scaled correctly! I'll figure it out on the weekend.

(You only get kerning if you turn it on in <kbd>Shift</kbd>+<kbd>F8</kbd>,<kbd>4</kbd>,<kbd>6</kbd> - so just don't do that for now 😄 )

**Comment by taviso** [comment]
Nevermind, I figured it out. I think it's looking pretty good:

![image](https://user-images.githubusercontent.com/123814/184068724-46c61954-27b4-48d4-bf6a-19f2e2f35121.png)

![image](https://user-images.githubusercontent.com/123814/184068741-83d7e6a6-fde6-45c8-a1c9-3e55a47ab01d.png)

Updated gszript.all:

[gscript.zip](https://github.com/taviso/wpunix/files/9306247/gscript.zip)

Please let me know what still needs to be fixed, I think it's getting pretty close to being good enough to release.

**Comment by emendelson** [comment]
I was just here updating an earlier comment and saw this. Will try later. About the slant - that's just an arbitrary number that WP uses to identify the typeface, so that it can try to match a font on a different printer to the font used in the machine that created the file. Typically it's easiest to copy the slant from an existing typeface. (The help file says the slant a signed value specifying the distance the top of a capital letter is moved from upright, in 1200ths of an inch.)

**Comment by emendelson** [comment]
Just tested the latest gscript.all, and it's amazingly good. The only issue I found was this:

If you select the Ghostscript font names (like Palladio and Nimbus No9 and Nimbus Sans), WP can't load them, and Courier prints instead. It turns out that the PFB  files have different PostScript names from the ones that (I would guess) are in the AFM files, so that the driver needs to be modified so that (for example) the Palladio Roman font gets designated "P052-Roman", and Nimbus Sans needs to be "NimbusSans-Regular". I made a couple of these substitutions by hand, and they worked.

Also, this is something I need to look into further: I suspect that the driver should hide the Ghostscript names (Nimbus, etc.), and simply change the Times font in the driver so that it uses the typeface, character map, and spacing table of NimbusNo9. Users expect Times, and this might make things smoother. But I'm not certain of this.

Here are two PDFs. I made the first after fixing the PostScript name for Palladio but not the other two that show up as Courier. The second shows the effect of the other PostScript name fixes:

[pp_0000000690_20__e-job_33.pdf](https://github.com/taviso/wpunix/files/9309037/pp_0000000690_20__e-job_33.pdf)

[pp_0000000788_21__e-job_34.pdf](https://github.com/taviso/wpunix/files/9309039/pp_0000000788_21__e-job_34.pdf)

**Comment by taviso** [comment]
The name given to GhostScript is the "load string", and for Palladio it's set to "URWPalladioL-Roma", which seems correct:

```
GS>/URWPalladioL-Roma findfont
Loading URWPalladioL-Roma font from /usr/share/fonts/type1/gsfonts/p052003l.pfb... 4390968 2862304 2752200 1455581 1 done.
```

As far as I know, all the other strings are just for the user and WordPerfect, so we can set them to anything at all, so no problem calling them "Times" and so on.

I guess I'm not sure where "P052-Roman" comes from, where do you see that string?

The names that ghostscript knows are configurable with your fontmap, it might be in `/etc/ghostscript/fontmap.d/10gsfonts.conf`.

**Comment by emendelson** [comment]
I've been testing in WSL Ubuntu, mostly because it's faster, so all this may apply only there, not in "real" Linux.

Look in the Ghostscript source code, under /Resource/Font - unless I'm completely wrong, these are the PFB files that Ghostscript uses when converting PostScript code to PDF. (I seem to remember that they get baked into the gs executable, but I'm sure of that only under Windows and macOS.) The first 14 bytes of the file P052-Roman are:

`%!PS-AdobeFont-1.0: P052-Roman`

Am I completely mistaken in thinking that the string that follows the colon (up to the space) is the PostScript name that needs to be in the driver?

Of course you're right that the names are configurable, but, unless I'm totally wrong (which happens all too often), it would be easier to adapt to the font names in the file than modifying the fontmap.

On my WSL Ubuntu setup, there is nothing in `/etc/ghostscript/fontmap.d` 

(I have the Ghostscript source code because I build my own executables that include some fonts that I use with WordPerfect.)

**Comment by taviso** [comment]
Ah, those are the URW fonts (the ones for level2 compatibility), not the base fonts, I haven't generated those ones yet, but it's easy to add them.

I guess I'm surprised you *only* have the level 2 fonts, I thought it was always optional, but maybe it's a compilation option. It's no problem, I'll start including those ones too.

Tavis.

**Comment by emendelson** [comment]
See my edit to the post that you were responding to. You've pointed toward the explanation...

**Comment by taviso** [comment]
I haven't tested it yet, but I quickly just generated maps for all the URW fonts, you're welcome to see if they work better on WSL!

[gscript.zip](https://github.com/taviso/wpunix/files/9310483/gscript.zip)

**Comment by emendelson** [comment]
Hmm... This (downloaded your fixed version) doesn't seem to be working in WSL. Notice the wrong euro character in the New Century Schoolbook line.

[pp_0000002354_30__e-job_43.pdf](https://github.com/taviso/wpunix/files/9310768/pp_0000002354_30__e-job_43.pdf)

**Comment by taviso** [comment]
Ah, I left in the original postscript fonts - I was planning to make them "fallback" fonts, although I don't know if that's really necessary! I guess I could mark them hidden.

**Comment by emendelson** [comment]
I think the solution is to remove the Ghostscript font names and reassign the original Postscript names (which everyone knows) to the character map and spacing tables of the Ghostscript fonts. I've done that for a half-dozen fonts here:

[GSCRIPT1.zip](https://github.com/taviso/wpunix/files/9311083/GSCRIPT1.zip)

And this is what the result looks like (which I think is right):

[pp_0000002537_32__e-job_45.pdf](https://github.com/taviso/wpunix/files/9311084/pp_0000002537_32__e-job_45.pdf)

No one knows what the URW font names mean, but everyone who's likely to use WP for printing will know the original PS names. Again, I only changed the two tables for the roman variety of a couple of the fonts (and I'm not sure I got the AvantGarde matching right). But let me know what you think...
